### PR TITLE
Make sure not to close the evaluator _ws_thread when done is not set yet

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
(Maybe) Resolves https://github.com/equinor/ert/issues/7642

**Approach**
Make sure to wait until done is set.

- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/ensemble_evaluator/evaluator.py [hunk 1]
```diff
@@ -144,6 +145,8 @@ def _cancelled_handler(self, events: List[CloudEvent]) -> None:
             )
             send_future.result()
             self._loop.call_soon_threadsafe(self._stop)
+            while not self._done.done():
+                time.sleep(0.1)
```
Comment (jonathan-eq) [inline src/ert/ensemble_evaluator/evaluator.py:149]: As this is repeated three times; should we put it in a separate method called `wait_until_done` or `sleep_until_done`?

Comment (xjules) [reply src/ert/ensemble_evaluator/evaluator.py:149]: Yes, we should.

**Comment by berland** [comment]
With this PR, the test

`fix_loop$ pytest tests/unit_tests/ensemble_evaluator/test_ensemble_legacy.py -k run_and_cancel_legacy_ensemble[using_sch --count=10000 -n 48`

passed. Looks good! 👍🏻

File: src/ert/ensemble_evaluator/evaluator.py (hunk 1 above)
Comment (xjules) [reply src/ert/ensemble_evaluator/evaluator.py]: I have replaced the `time.sleep` with `threading.Event()`

**Comment by eivindjahren** [comment]
> With this PR, the test
> 
> `fix_loop$ pytest tests/unit_tests/ensemble_evaluator/test_ensemble_legacy.py -k run_and_cancel_legacy_ensemble[using_sch --count=10000 -n 48`
> 
> passed. Looks good! 👍🏻

I played a bit with the repeat numbers and the failing behavior becomes consistent at about ~20 repeats. Each time it takes about 4 seconds to run the test so I think it seems reasonable to add
```
@pytest.repeat(20)
```
to the test so that we get consistent test results rather than flaky behavior. We run with pytest-xdist anyways so this should take neglible additional time (at least in comparison to the `using_scheduler` fixture). This also means adding `pytest-repeat` to the dev dependencies in `pyproject.toml`.

**Comment by eivindjahren** [comment]
We talked about the repeat idea, and came up with the idea of just parametrizing the test:
```python
@pytest.mark.timeout(60)
@pytest.mark.usefixtures("using_scheduler")
@pytest.mark.parametrize("num_reals, num_jobs", product(range(2, 6), range(2, 6)))
def test_run_and_cancel_legacy_ensemble(
    tmpdir, make_ensemble_builder, monkeypatch, num_reals, num_jobs
):
    print(num_reals, num_jobs)
    custom_port_range = range(1024, 65535)
    with tmpdir.as_cwd():
        ensemble = make_ensemble_builder(
            monkeypatch, tmpdir, num_reals, num_jobs, job_sleep=40, max_runtime=1
        ).build()
        config = EvaluatorServerConfig(
            custom_port_range=custom_port_range,
            custom_host="127.0.0.1",
            use_token=False,
            generate_cert=False,
        )

        evaluator = EnsembleEvaluator(ensemble, config, 0)

        evaluator.start_running()

        async def _run_monitor():
            async with Monitor(config) as mon:
                cancel = True
                with contextlib.suppress(
                    ConnectionClosed
                ):  # monitor throws some variant of CC if dispatcher dies
                    async for _ in mon.track():
                        # Cancel the ensemble upon the arrival of the first event
                        if cancel:
                            await mon.signal_cancel()
                            cancel = False

        run_monitor_in_loop(_run_monitor)
        assert evaluator._ensemble.status == state.ENSEMBLE_STATE_CANCELLED

        # realisations should not finish, thus not creating a status-file
        for i in range(num_reals):
            assert not os.path.isfile(f"real_{i}/status.txt")
```

**Comment by xjules** [comment]
Closing this PR as this one is not required anymore it seems. Relates https://github.com/equinor/ert/pull/8000

**Pull request closed without merging**